File: backend/src/main/java/com/programmers/pcquotation/domain/estimaterequest/entity/EstimateRequest.java
package com.programmers.pcquotation.domain.estimaterequest.entity;

import java.time.LocalDateTime;
import java.util.List;

import com.programmers.pcquotation.domain.customer.entity.Customer;

import com.programmers.pcquotation.domain.estimate.entity.Estimate;
import com.programmers.pcquotation.domain.estimaterequest.dto.EstimateRequestData;
import jakarta.persistence.*;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Getter;
import lombok.NoArgsConstructor;

@Entity
@Getter
@Builder
@NoArgsConstructor
@AllArgsConstructor
public class EstimateRequest {
	@Id
	@GeneratedValue(strategy = GenerationType.IDENTITY)
	private Integer id;

	@Column(length = 20)
	private String purpose;

	@Column(columnDefinition = "INTEGER")
	private Integer budget;

	@Column(length = 200)
	private String otherRequest;

	LocalDateTime createDate;

	@ManyToOne
	private Customer customer;

	@OneToMany(mappedBy = "estimateRequest", cascade = CascadeType.REMOVE)
	private List<Estimate> estimate;

	@Enumerated(EnumType.STRING)
	private EstimateRequestStatus status; // 0: 대기 중, 1: 채택됨

	public void UpdateEstimateRequest(EstimateRequestData estimateRequestData) {
		this.purpose = estimateRequestData.purpose();
		this.budget = estimateRequestData.budget();
		this.otherRequest = estimateRequestData.otherRequest();
	}

	public void UpdateDeliveryStatus(EstimateRequestStatus estimateRequestStatus) {
		this.status = estimateRequestStatus;
	}
}

File: backend/src/main/java/com/programmers/pcquotation/domain/estimaterequest/entity/EstimateRequest.kt
package com.programmers.pcquotation.domain.estimaterequest.entity

import com.programmers.pcquotation.domain.customer.entity.Customer
import com.programmers.pcquotation.domain.estimate.entity.Estimate
import com.programmers.pcquotation.domain.estimaterequest.dto.EstimateRequestData
import jakarta.persistence.*
import org.springframework.data.annotation.CreatedDate
import java.time.LocalDateTime

@Entity
class EstimateRequest @JvmOverloads constructor(
    @Column(length = 20)
    var purpose: String,

    @Column(columnDefinition = "INTEGER")
    var budget: Int,

    @Column(length = 200)
    var otherRequest: String,

    @ManyToOne
    var customer: Customer,
) {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    var id: Int = 0

    @CreatedDate
    var createDate: LocalDateTime = LocalDateTime.now()

    @Enumerated(EnumType.STRING)
    var status: EstimateRequestStatus = EstimateRequestStatus.WAIT // 0: 대기 중, 1: 채택됨

    @OneToMany(mappedBy = "estimateRequest", cascade = [CascadeType.REMOVE])
    var estimate: MutableList<Estimate>? = null

    fun updateEstimateRequest(estimateRequestData: EstimateRequestData) {
        this.purpose = estimateRequestData.purpose
        this.budget = estimateRequestData.budget
        this.otherRequest = estimateRequestData.otherRequest
    }

    fun updateEstimateRequestStatus(estimateRequestStatus: EstimateRequestStatus) {
        this.status = estimateRequestStatus
    }

    constructor(estimateRequestData: EstimateRequestData, customer: Customer) : this(
        purpose = estimateRequestData.purpose,
        budget = estimateRequestData.budget,
        otherRequest = estimateRequestData.otherRequest,
        customer = customer
    )

}